woocommerce/woocommerce-gateway-stripe
# Fix customer deletion and payment detachment in staging site

Fixes #2614 

When a user is deleted from a staging site, payment method tokens are getting detached from customers on Stripe's side. This hampers the customer's account and subscriptions on the production site. The expected behavior is not to detach the payment method from the Stripe customer if the user is deleted from a staging site. If the payment method deletion takes place by customer interaction (from `My account` page of the customer), then the method should be detached for staging or production any site.

## Changes proposed in this Pull Request:
- Added `should_detach_payment_method_from_customer` function to check if we should make the detach payment method request to Stripe.
- This function returns false if the site is a staging site and the request is coming from the admin dashboard (which is mainly the delete user request). Used WP method `wp_get_environment_type` to determine the environment.
- This function returns true for all other cases and allows sending a detach payment method request to Stripe.

## Testing instructions
- For the sake of testing in the local environment, you may replace `staging` with `development` in the condition check [here](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2672/files#diff-9b5ad4eb8d7bffb53e706353d9fbc116c6d4f6c8f328eb154510ed24fe351b5aR370).
- Purchase a subscription with a new user `U`.
- With this new user go to the **My account -> Payment methods** page and add 3 more payment methods there.
- Find the corresponding customer in your Stripe dashboard.
- Ensure that the customer has a total of 4 payment methods attached.
- Now delete a payment method from **My account -> Payment methods** page (except the one used for subscription)
- Check the customer in your Stripe dashboard again and ensure that it has 3 payment methods attached now. As the payment method was deleted directly by the customer's interaction, we should detach the payment method here.
- Now go to **Users > All Users** in the `wp-admin` as an admin user.
- Find user `U` and delete the user.
- Check the customer in your Stripe dashboard again and ensure that it still has the 3 payment methods attached.
